### ISSUE TITLE
Think whether we want to keep the `IsResource` wrapping stuff

### ISSUE THREAD

**Issue by arianvp** [wireapp/servant-uverb#11]
I find it rather akward to use. 

TODO. Write up the design rationale and the pros and cons

**Comment by fisx** [comment]
I agree that what we're currently doing is nice, but I really want to avoid having to set the status code in the handler.  It *has* to be set in the type for swagger2 to find it, and the user shouldn't be forced to set it twice.

As long as we can find a way to accomplish this, I'm happy.

Let me try again:

```haskell
data WithStatus (k :: Nat) a = WithStatus a

type API = Get ... (WithStatus Bool)

handler = pure $ WithStatus True
```

Is it really that easy?  What did I miss when I made up the convoluted `IsResource` thing?  I think I'm missing something now...

**Comment by arianvp** [comment]
GHC complains about an ambigious type variable.

```
pure $ WithStatus @200 Bool
```
fixes it ,but why do we need the annotation in the first place?

**Comment by fisx** [comment]
we don't, though!  this compiles fine with ghc 8.4.3:

```haskell
data WithStatus' (k :: Nat) a = WithStatus' a

type API' = Get '[JSON] (WithStatus' 500 Bool)

handler' :: Server API'
handler' = pure $ WithStatus' True
```

so what was our problem?

**Comment by arianvp** [comment]
Because that's just one!  We have an open union of `WithStatus`es

**Comment by arianvp** [comment]
How about an open type family?

```haskell
type family HasStatus (x :: *) :: Nat


type instance HasStatus Bool 500 
```

But then we run into the orphan instance problem agian,

Anyhow, I think a type family here is more elegant than the typeclass we have.

**Comment by fisx** [comment]
...  and we could avoid the orphans the same way we avoided them with the `Resource` newtype that i suggested be defined inside the app using uverb.

**Comment by alpmestan** [comment]
> we don't, though! this compiles fine with ghc 8.4.3:
> 
> ```haskell
> data WithStatus' (k :: Nat) a = WithStatus' a
> 
> type API' = Get '[JSON] (WithStatus' 500 Bool)
> 
> handler' :: Server API'
> handler' = pure $ WithStatus' True
> ```
> 
> so what was our problem?

You added the type sig for `handler'`, which told GHC what status code that `WithStatus' True` was associated to.

**Comment by arianvp** [comment]
@alpmestan the actualy problem is this:

```
data WithStatus' (k :: Nat) a = WithStatus' a

type API' = Get '[JSON] '[WithStatus' 500 Bool, WithStatus '403 Bool]

handler' :: Server API'  -- ~  IO (NS I '[WithStatus 500 Bool, WithStatus 403 Bool])
handler' = pure $ injectNS $ WithStatus' True
```

Now `k` is ambigious


You need instead:

```
handler' :: Server API'  -- ~  IO (NS I '[WithStatus 500 Bool, WithStatus 403 String])
handler' = pure $ injectNS $ WithStatus' @500 True
```

**Comment by alpmestan** [comment]
Right, the ambiguity "requires" 2 different status codes for `Bool`. Anyway, sorry for the noise, I'll keep following silently. =)

**Comment by fisx** [comment]
I think we've settled for a way out in #15
